File: tests/fixtures/positive/exceptions/InstanceOfCheckForException.java
package com.example.exceptions;

import java.io.IOException;
import java.net.SocketException;

class ErrorClassifier {

  // instanceof type-check on a value that is NOT the caught variable is the
  // "instanceof instead of polymorphism" smell this rule targets.
  String classify(Throwable failure) {
    try {
      return "ok";
    } catch (Exception outer) {
      if (failure instanceof SocketException) {
        return "socket";
      }
      return "other";
    }
  }
}

File: tests/fixtures/positive/exceptions/InstanceOfCheckForException.kt
package com.example.exceptions

import java.io.IOException

class ErrorClassifier {

    // Type-checking an exception value that is NOT a caught variable to
    // branch on its concrete type is the "instanceof instead of
    // polymorphism" smell this rule targets.
    fun classify(failure: Throwable): String {
        try {
            return "ok"
        } catch (outer: Exception) {
            // `failure` is a parameter, not the caught variable `outer`.
            if (failure is IOException) {
                return "io"
            }
            return "other"
        }
    }
}
